protocols/kad/src/jobs.rs

```diff
@@ -0,0 +1,363 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::record::{Record, ProviderRecord, store::RecordStore};
+
+use libp2p_core::PeerId;
+use futures::prelude::*;
+use multihash::Multihash;
+use std::collections::HashSet;
+use std::time::Duration;
+use std::vec;
+use wasm_timer::{Instant, Delay};
+
+/// The maximum number of queries towards which background jobs
+/// are allowed to start new queries on an invocation of
+/// `Kademlia::poll`.
+pub const JOBS_MAX_QUERIES: usize = 100;
+
+/// The maximum number of new queries started by a background job
+/// per invocation of `Kademlia::poll`.
+pub const JOBS_MAX_NEW_QUERIES: usize = 10;
+
+/// A background job run periodically.
+#[derive(Debug)]
+struct PeriodicJob<T> {
+    interval: Duration,
+    state: PeriodicJobState<T>,
+}
+
+impl<T> PeriodicJob<T> {
+    fn is_running(&self) -> bool {
+        match self.state {
+            PeriodicJobState::Running(..) => true,
+            PeriodicJobState::Waiting(..) => false,
+        }
+    }
+
+    /// Cuts short the remaining delay, if the job is currently waiting
+    /// for the delay to expire.
+    fn asap(&mut self) {
+        if let PeriodicJobState::Waiting(delay) = &mut self.state {
+            delay.reset(Instant::now() - Duration::from_secs(1))
+        }
+    }
+
+    /// Returns `true` if the job is currently not running but ready
+    /// to be run, `false` otherwise.
+    fn is_ready(&mut self, now: Instant) -> bool {
+        if let PeriodicJobState::Waiting(delay) = &mut self.state {
+            if now >= delay.deadline() || delay.poll().map(|a| a.is_ready()).unwrap_or(false) {
+                return true
+            }
+        }
+        false
+    }
+}
+
+/// The state of a background job run periodically.
+#[derive(Debug)]
+enum PeriodicJobState<T> {
+    Running(T),
+    Waiting(Delay)
```

Comment:
Why contains `PeriodicJobState::Waiting` a future instead of a deadline value?

Reply:
The idea is that when a job is polled (see `PutRecordJob::poll` and `AddProviderJob::poll`) and returns `NotReady` the current task is woken up when the job is ready to run, since there may be nothing else going on otherwise. This polling of the jobs is done in `Kademlia::poll`. Does that make sense or did you have something else in mind with your question?